File: day01/solution.ts
import { readLines } from "https://deno.land/std@0.172.0/io/mod.ts";

const part1 = async () => {
  const file = await Deno.open("input.txt");
  let maxCaloriesPerElf = 0;
  let maxCalories = 0;
  for await (const line of readLines(file)) {
    if (line === "") {
      maxCalories = Math.max(maxCalories, maxCaloriesPerElf);
      maxCaloriesPerElf = 0;
      continue;
    }
    maxCaloriesPerElf += parseInt(line);
  }
  console.log(maxCalories);
};

const part2 = async () => {
  const file = await Deno.open("input.txt");
  let maxCaloriesPerElf = 0;
  const maxCalories: number[] = [0, 0, 0];
  for await (const line of readLines(file)) {
    if (line === "") {
      const minIndex = maxCalories.indexOf(Math.min(...maxCalories));
      maxCalories[minIndex] = Math.max(
        maxCaloriesPerElf,
        maxCalories[minIndex]
      );
      maxCaloriesPerElf = 0;
      continue;
    }
    maxCaloriesPerElf += parseInt(line);
  }
  console.log(maxCalories.reduce((a, b) => a + b, 0));
};

await part1();
await part2();

File: day01/solution.py
import functools


def part1():
    lines = []
    max_calories = 0
    max_calories_per_elf = 0
    with open("input.txt") as f:
        lines = f.readlines()
    for line in lines:
        if line == "\n":
            max_calories = max(max_calories, max_calories_per_elf)
            max_calories_per_elf = 0
            continue
        max_calories_per_elf += int(line.split()[0])

    print(max_calories)


def part2():
    lines = []
    max_calories = [0, 0, 0]
    max_calories_per_elf = 0
    with open("input.txt") as f:
        lines = f.readlines()
    for line in lines:
        if line == "\n":
            min_index = max_calories.index(min(max_calories))
            max_calories[min_index] = max(max_calories[min_index], max_calories_per_elf)
            max_calories_per_elf = 0
            continue
        max_calories_per_elf += int(line.split()[0])

    print(functools.reduce(lambda x, y: x + y, max_calories))


if __name__ == "__main__":
    part1()
    part2()
